components/net/local_directory_listing.rs

```diff
@@ -0,0 +1,388 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+use std::collections::BTreeMap;
+use std::ffi::OsString;
+use std::io;
+use std::path::PathBuf;
+
+use embedder_traits::resources::{read_string, Resource};
+use headers::{ContentType, HeaderMapExt};
+use net_traits::request::Request;
+use net_traits::response::{Response, ResponseBody};
+use net_traits::{NetworkError, ResourceFetchTiming};
+use servo_config::pref;
+use servo_url::ServoUrl;
+use time_03::OffsetDateTime;
+use url::Url;
+
+pub fn is_request_allowed(request: &Request) -> Result<(), Response> {
+    if !pref!(network.local_directory_listing.enabled) {
+        // If you want to be able to browse local directories, configure Servo prefs so that
+        // "network.local_directory_listing.enabled" is set to true.
+        Err(Response::network_error(NetworkError::Internal(
+            "Local directory listing feature has not been enabled in preferences".into(),
+        )))
+    } else if !request.origin.is_opaque() {
+        // Checking for an opaque origin as a shorthand for user activation
+        // as opposed to a request originating from a script.
+        // TODO: Raise and link to GitHub issue to request consideration of this concern.
+        Err(Response::network_error(NetworkError::Internal(
+            "Cannot request local directory listing from non-local origin.".to_string(),
+        )))
+    } else {
+        Ok(())
+    }
+}
+
+pub fn fetch(
+    request: &mut Request,
+    url: ServoUrl,
+    directory: impl Into<DirectorySummary>,
+) -> Response {
+    let url = if !url.path().ends_with('/') {
+        // Re-read the path as a directory, so that Servo adds the
+        // forward-slash to the end of the URL (at least internally,
+        // though the URL bar does not reflect this) and loads the
+        // linked directories and files correctly when clicked.
+        if let Ok(dir_url) = Url::from_directory_path(url.path()) {
+            ServoUrl::from_url(dir_url)
+        } else {
+            return Response::network_error(NetworkError::Internal(format!(
+                "Unable to parse local directory path {}",
+                url
+            )));
+        }
+    } else {
+        url
+    };
+    let mut response = Response::new(url, ResourceFetchTiming::new(request.timing_type()));
+    response.headers.typed_insert(ContentType::html());
+    let directory_summary = directory.into();
```

Comment:
I always prefer to see a named container instead of a gaggle of method/function parameters. Most "clean code" books agree that named containers are safer/clearer than a sequence of loose method/function parameters, though I know some people disagree with this view.

Are you suggesting we get rid of the `DirectorySummary` struct altogether, or just that we break it open (to pull out the inner values) before calling `build_html_directory_listing`? If we get rid of it altogether, then the `From<PathBuf>` trait has to be binned too, and all of the unit tests which currently mock-up example imaginary directories by populating `DirectorySummary` objects.

Reply:
Okay, reworked things to get rid of the `DirectorySummary` container.